src/goto-instrument/code_contracts.cpp

```diff
@@ -209,14 +257,47 @@ void code_contractst::check_apply_invariants(
     converter.goto_convert(assertion, havoc_code, mode);
     havoc_code.instructions.back().source_location.set_comment(
       "Check that loop invariant is preserved");
-    auto offset = havoc_code.instructions.size();
-    goto_function.body.insert_before_swap(loop_end, havoc_code);
-    std::advance(loop_end, offset);
   }
 
+  if(!variant.is_nil())
+  {
+    // Generate: a declaration of a temporary variable that stores the variant
+    // after one arbitrary iteration of the loop
+    havoc_code.add(
+      goto_programt::make_decl(new_variant_symbol, loop_head->source_location));
+
+    // Generate: an assignment to store the loop variant after one iteration of
+    // the loop
+    source_locationt declaration_new_variant_source_location{
+      loop_head->source_location};
+    havoc_code.add(goto_programt::make_assignment(
+      new_variant_symbol, variant, declaration_new_variant_source_location));
+    havoc_code.instructions.back().source_location.set_comment(
+      "Evaluate and store the loop variant after the loop");
+
+    // Generate: assertion that the loop variant after the loop is smaller than
+    // the loop variant before the loop
+    source_locationt assignment_source_location{loop_head->source_location};
+    havoc_code.add(goto_programt::make_assertion(
+      binary_relation_exprt(new_variant_symbol, ID_lt, old_variant_symbol),
+      assignment_source_location));
```

Comment:
Instead of generating a goto instruction directly, we should be calling `goto_convert` to properly translate a high-level assertion to appropriate goto instructions. Use:

```c
    code_assertt variant_assertion{binary_relation_exprt(new_variant_symbol, ID_lt, old_variant_symbol)};
    variant_assertion.add_source_location() = loop_end->source_location;
    converter.goto_convert(variant_assertion, havoc_code, mode);
```

Reply:
Thanks a lot for the code.